### PR TITLE
conditiona_tool_implementation

### PR DESCRIPTION
**Pull request by SahilHeartzai**

### CODE REVIEW

File: langchain.log [hunk 1]
```diff
@@ -1,420 +1,62 @@
...
-2024-02-16 11:25:04,163 - langchain_gpt_api- [RequestID: None] - INFO - memory object is not none and metadata is {'request_Id': None}
-2024-02-16 11:25:04,181 - langchain_gpt_api- [RequestID: None] - INFO - After: memory object is not none and metadata is {'request_Id': None}
-2024-02-16 11:25:09,083 - langchain_gpt_api- [RequestID: a91abecf-e36d-4014-adac-150e2d22deed] - INFO - query------> explain this image
-2024-02-16 11:25:09,332 - langchain_gpt_api- [RequestID: a91abecf-e36d-4014-adac-150e2d22deed] - INFO - calling for 1 times
-2024-02-16 11:25:09,333 - langchain_gpt_api- [RequestID: a91abecf-e36d-4014-adac-150e2d22deed] - INFO - len---->2952
-2024-02-16 11:25:09,338 - langchain_gpt_api- [RequestID: a91abecf-e36d-4014-adac-150e2d22deed] - INFO - len---->5180
-2024-02-16 11:25:12,154 - langchain_gpt_api- [RequestID: a91abecf-e36d-4014-adac-150e2d22deed] - INFO - hellpppppppppppppppp-->
+2024-02-24 18:31:57,445 - langchain_gpt_api- [RequestID: 10000000001] - INFO - before: memory object is not none and metadata is {'request_Id': 10000000001}, False
```
Comment (lawlietlight) [inline langchain.log]: Please remove non code files which are generated

File: langchain_gpt_api.py [hunk 2]
```diff
@@ -1083,7 +1289,47 @@ def chat():
     data = request.get_json()
     user_id = data.get('user_id', None)
     request_id = data.get('request_id', None)
+    req_tool = data.get('tools', None)
+    tools_dict = {1:'google_search', 2:'Calculator', 3:'OpenAPI_Specification', 4:'FULL_HISTORY', 5:'Text to image', 6:'Image_Inference_Tool', 7:'Data_Extraction_From_URL', 8:'User_details_tool'}
+    tool_desc = {
+        'google_search': '''Search Google for recent results and retrieve URLs that are suitable for web crawling. Ensure that the search responses include the source URL from which the data was extracted. Always present this URL in the response as an HTML anchor tag. This approach ensures clear attribution and easy navigation to the original source for each piece of extracted information. Give urls for the source''',
```
Comment (lawlietlight) [inline langchain_gpt_api.py]: Can't we define this at class level only once and extract and use in both places

**Review by lawlietlight — CHANGES_REQUESTED** [review]

File: langchain.log (hunk 1 above)
Comment (SahilHeartzai) [reply langchain.log]: done

File: langchain_gpt_api.py (hunk 2 above)
Comment (SahilHeartzai) [reply langchain_gpt_api.py]: done

File: req.txt
Comment (lawlietlight) [inline req.txt:1]: how many requirements.txt file exists? why is this file name in non standard format?

**Review by lawlietlight — CHANGES_REQUESTED** [review]

File: req.txt
Comment (lawlietlight) [reply req.txt:1]: is this addressed?

**Pull request merged**